From bddcfef51b63925ad20f6364d44dcf9bd3ba1289 Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Date: Fri, 25 Sep 2020 14:03:02 +0300
Subject: [PATCH 178/269] s32g: wkpu: Correct "reg" property interpretation

Use fdt_get_reg_props_by_index to read "reg" property. It will take
into consideration the number of cells used for address and size
specification.

Issue: ALB-4294, ALB-5510
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <Ghennadi.Procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/nxp/s32g/s32g_wkpu.c    | 7 ++++---
 plat/nxp/s32g/include/s32g_dt.h | 4 +++-
 plat/nxp/s32g/platform.mk       | 1 +
 3 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/nxp/s32g/s32g_wkpu.c b/drivers/nxp/s32g/s32g_wkpu.c
index 3173ff0b7..4b103fae0 100644
--- a/drivers/nxp/s32g/s32g_wkpu.c
+++ b/drivers/nxp/s32g/s32g_wkpu.c
@@ -3,6 +3,7 @@
  * Copyright 2020 NXP
  */
 #include <common/debug.h>
+#include <common/fdt_wrappers.h>
 #include <dt-bindings/reset/s32gen1-wkpu.h>
 #include <errno.h>
 #include <lib/mmio.h>
@@ -82,7 +83,7 @@ static void init_wkpu(struct s32gen1_wkpu *wkpu)
 
 static int init_from_dt(void *fdt, int fdt_offset, struct s32gen1_wkpu *wkpu)
 {
-	const fdt32_t *irq_ptr, *reg;
+	const fdt32_t *irq_ptr;
 	uint32_t irq_num;
 	uint32_t pull;
 	int len;
@@ -94,7 +95,7 @@ static int init_from_dt(void *fdt, int fdt_offset, struct s32gen1_wkpu *wkpu)
 	if (wkpu->dt_info.status != DT_ENABLED)
 		return -1;
 
-	reg = fdt_getprop(fdt, fdt_offset, "reg", &len);
+	(void) fdt_getprop(fdt, fdt_offset, "reg", &len);
 	/* WKPU & GPR ranges */
 	if (len < 4 * sizeof(uint32_t)) {
 		ERROR("Missing GPR registers\n");
@@ -102,7 +103,7 @@ static int init_from_dt(void *fdt, int fdt_offset, struct s32gen1_wkpu *wkpu)
 	}
 
 	/* GPR Base address */
-	wkpu->gpr = fdt32_to_cpu(reg[2]);
+	(void) fdt_get_reg_props_by_index(fdt, fdt_offset, 1, &wkpu->gpr, NULL);
 
 	irq_ptr = fdt_getprop(fdt, fdt_offset, "nxp,irqs", &len);
 	if (!irq_ptr) {
diff --git a/plat/nxp/s32g/include/s32g_dt.h b/plat/nxp/s32g/include/s32g_dt.h
index 909f10c9d..1de93b458 100644
--- a/plat/nxp/s32g/include/s32g_dt.h
+++ b/plat/nxp/s32g/include/s32g_dt.h
@@ -7,11 +7,13 @@
 #ifndef S32G_DT_H
 #define S32G_DT_H
 
+#include <stdint.h>
+
 #define DT_DISABLED	0
 #define DT_ENABLED	1
 
 struct dt_node_info {
-	uint32_t base;
+	uintptr_t base;
 	int32_t clock;
 	int32_t reset;
 	uint32_t status;
diff --git a/plat/nxp/s32g/platform.mk b/plat/nxp/s32g/platform.mk
index 8cc37945a..ff0b08d55 100644
--- a/plat/nxp/s32g/platform.mk
+++ b/plat/nxp/s32g/platform.mk
@@ -42,6 +42,7 @@ PLAT_BL_COMMON_SOURCES	+= plat/nxp/s32g/s32g_lowlevel_common.S \
 			   drivers/delay_timer/generic_delay_timer.c \
 			   drivers/nxp/s32g/memory_pool.c \
 			   lib/cpus/aarch64/cortex_a53.S\
+			   common/fdt_wrappers.c \
 			   ${GICV3_SOURCES} \
 			   ${BL31SRAM_SRC_DUMP} \
 
-- 
2.17.1

